upcoming: [M3-9495] - Disable APL for LKE-E clusters

mjac0bs opened the pull request:
## Description 📝

APL won't support LKE-E clusters until later this year (currently roadmapped for Q3). We need to disable it for LKE-E clusters until then and since it's upcoming let users know it is 'coming soon'.

## Changes  🔄

- Created a new variable `isAPLSupported` to check `showAPL` **and** the `selectedTier`
- Updated logic in ApplicationPlatform to:
   - Add 'COMING SOON' to chip for LKE-E
   - Disable both radio buttons
   - Check the 'No' button
- Updated `lke-create.spec.ts` to check that an LKE-E cluster tier selection results in the section being disabled
- Added `ApplicationPlatform.test.tsx`

@hasyed-akamai's PR #11581 will need to replicate these changes for parity in the RHF cluster create flow.

## Target release date 🗓️

3/25

## Preview 📷

In summary, this was the change:
| Conditions | State of APL section | Before  | After  |
| ----------- | --------------- | ------- | ------- |
| LKE-E feature flag enabled, LKE-E tier selected, enrolled in APL beta | APL section visible and disabled | ![Screenshot 2025-03-11 at 11 33 48 PM](https://github.com/user-attachments/assets/22474857-a57e-4f2f-aafb-5ea43ee4ca03) | ![LKE-E-Flag-Enabled-Enrolled-in-Beta-2](https://github.com/user-attachments/assets/7346a14c-c390-426e-b177-8f9aae08aa8e) |

A more comprehensive reference of all the other possible states:
| Conditions | State of APL section |  Screenshot |
| -----------  | --------------- | ------- |
| LKE-E feature flag disabled, not enrolled in APL beta | APL section hidden  | ![LKE-E-Flag-Disabled-Not-Enrolled-in-Beta](https://github.com/user-attachments/assets/6b62e86c-1b01-44d9-b99e-5ca66ec8a431) |
| LKE-E feature flag disabled, enrolled in APL beta | APL section visible and accessible | ![LKE-E-Flag-Disabled-Enrolled-in-Beta](https://github.com/user-attachments/assets/9b0db23a-d249-4b7b-95aa-1aa2aa1a69cf) |
| LKE-E feature flag enabled, not enrolled in APL beta | APL section hidden | ![LKE-E-Flag-Enabled-Not-Enrolled-in-Beta-2](https://github.com/user-attachments/assets/16533f97-6396-4767-9ce6-1418aef9c5f8) ![LKE-E-Flag-Enabled-Not-Enrolled-in-Beta-1](https://github.com/user-attachments/assets/7266fada-d994-484f-bb47-d3bcb0c0f90d) |
| LKE-E feature flag enabled, LKE tier selected, enrolled in APL beta | APL section visible and accessible  | ![LKE-E-Flag-Enabled-Enrolled-in-Beta-1](https://github.com/user-attachments/assets/7bc42a42-c224-46e2-b695-3f9b0c665e29) |


## How to test 🧪

### Prerequisites

(How to setup test environment)

- **Be enrolled in the Akamai App Platform beta** (https://cloud.linode.com/betas)
- Have the LKE-E feature flag enabled and customer tag on your account (see project tracker)

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] With the Beta enabled, select the LKE-E tier and observe the APL section is visible

### Verification steps

(How to verify changes)

- [ ] With the Beta enabled, select the LKE-E tier and observe the APL section is disabled, the No radio button is checked, and the chip includes 'COMING SOON'
- [ ] With the Beta enabled, select the LKE tier and observe the APL section is enabled, no radio button selection is checked, the the chip reads 'BETA' (aka: no changes from prod today)

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

jdamore-linode commented on packages/manager/cypress/e2e/core/kubernetes/lke-create.spec.ts:
I think we might also have to mock the APL feature flag and beta: I don't see APL shown at all when running this test regardless of whether LKE-E is selected, but these additions did the trick for me:

```typescript
      // Line ~1078
      mockAppendFeatureFlags({
        lkeEnterprise: { enabled: true, la: true },
        apl: { enabled: true },
      }).as('getFeatureFlags');
```

```typescript
      // Line ~1120, requires importing accountBetaFactory
      mockGetAccountBeta(accountBetaFactory.build({
        id: 'apl',
        label: 'Akamai App Platform Beta',
      })).as('getAccountBeta');
```

mjac0bs replied:
😅 Thanks, I wasn't thinking there. Will add.

jdamore-linode commented on packages/manager/cypress/e2e/core/kubernetes/lke-create.spec.ts:
```diff
@@ -1215,6 +1216,11 @@ describe('LKE Cluster Creation with LKE-E', () => {
         .should('be.enabled')
         .click();
 
+      // Confirm the APL section is not shown.
+      cy.findByTestId('apl-label').should('not.exist');
+      cy.findByTestId('apl-radio-button-yes').should('not.exist');
+      cy.findByTestId('apl-radio-button-no').should('not.exist');
+
```
Related to the above, can we additionally add assertions to confirm that the APL section is present when LKE-E isn't selected?

mjac0bs replied:
I believe [this test](https://github.com/linode/manager/blob/2dd117eda218eb1367cd90781186b2056b102e38/packages/manager/cypress/e2e/core/kubernetes/lke-create.spec.ts#L372) is [already accomplishing](https://github.com/linode/manager/blob/2dd117eda218eb1367cd90781186b2056b102e38/packages/manager/cypress/e2e/core/kubernetes/lke-create.spec.ts#L460-L465) that, right?

jdamore-linode replied:
Oh yeah, you're absolutely right. There is one distinction I'll call out which is that adding the assertions here gives us explicit coverage that the APL section is present _when the LKE-E feature is enabled but not selected during cluster creation_, whereas the other test doesn't take the state of LKE-E into account. So if there were ever a logic issue involving APL/LKE-E that resulted in APL being hidden, it would get caught by this test but may not get caught by the other.

(I have no idea what the logic for the cluster create page is like, so if this comes across as a farfetched hypothetical we can definitely leave it alone, just wanted to point it out in case it's an important distinction from a dev's POV)

mjac0bs replied:
I left the case of 'LKE-E feature enabled but LKE (standard) cluster tier selected' uncovered in tests because it is essentially the same flow as 'LKE-E feature disabled'. The code sets a default of `standard` for the `selectedTier`. The account beta endpoint determines whether or not the APL section is visible to users, and the section visible + the `selectedTier` determine whether the section is enabled or disabled.

Basically, if there were an issue with the LKE tier selection in the LKE-E flow, the first test could catch it.

jdamore-linode reviewed (approved):
Nice, thanks @mjac0bs!

mjac0bs commented:
> If the goal is to disable the panel, then that's what we should do, not hide/remove it, especially if this will come at a later time. 

Thanks, you're right that this is unclear. I'm following up with Product for clarity on this in #dev-apl.

> This form is already quite complex, and displaying elements conditionally could add confusion IMO.

To an extent, we're already doing this (displaying elements conditionally) with the HA form section being hidden once the LKE-E cluster is selected. This was accepted by Product since HA is mentioned as included in the cluster tier section. I see your point that this is a different situation where we'd potentially be "making the user wonder why this is happening and why this isn't possible (or will ever be)".

mjac0bs commented:
Reopening this up for review with updated UX treatment, which Product has agreed upon and the APL team is aware of.

The description is up to date.

hana-akamai commented on packages/manager/src/features/Kubernetes/CreateCluster/ApplicationPlatform.tsx:
```diff
@@ -49,19 +54,27 @@ export const ApplicationPlatform = (props: APLProps) => {
+          <Chip
+            color="primary"
+            data-testid="apl-beta-chip"
+            label={CHIP_COPY}
```
Can we make the chip text smaller?

| Medium  | Small   |
| ------- | ------- |
| ![image](https://github.com/user-attachments/assets/5855be0b-a74c-4c49-885c-a84b28a86408) | ![image](https://github.com/user-attachments/assets/2311cfa2-8db8-4d30-bc31-a2686e4a0b17) |

```suggestion
            label={CHIP_COPY}
            size="small"
```

mjac0bs replied:
Done in [b5362b1](https://github.com/linode/manager/pull/11809/commits/b5362b1d9da472c8d66f2deff64c9ba8c39aea3c). Took a look through our other chips and confirmed the majority of them had small text, so followed suit here. Thanks for the suggestion.

abailly-akamai reviewed (approved):
Looks great!

Confirmed new disabled treatment + chip depending on LKE-E ✅ 
Confirm inputs are properly disabled ✅

mjac0bs commented:
Fixed a bug and added a unit test that would catch it. A bad nullish coalescing kept the radio from being checked when the section was enabled, because the value was always `undefined`.

| Before | After |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/e63825fe-4790-4943-bc91-961e67939b0a"/> | <video src="https://github.com/user-attachments/assets/f59e9d64-03cd-4570-b1f7-d5b4561cfd82"/> | 

Will merge in the morning after checks pass.
